### PR TITLE
extract common functions from applyOverwriteDocument

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            new XrefInterpreter()
+        );
+
+        private static readonly SchemaProcessor XrefSpecUpdater = new SchemaProcessor(
+            new XrefPropertiesInterpreter()
+        );
+
+        private static readonly Merger Merger = new Merger();
+
+        public static void ReExportXrefSpec(FileModel fileModel, IHostService host, BaseSchema schema)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs]: naming: `Update`?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        public static object BuildOverwriteWithSchema(FileModel owModel, OverwriteDocumentModel overwrite, IHostService host, BaseSchema schema)
+        {
+            if (overwrite == null || owModel == null)
+            {
+                return null;
+            }
+
+            if (host == null)
+            {
+                throw new ArgumentException("host");
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs]: throw `ArgumentNullException`

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            dynamic overwriteObject = ConvertToObjectHelper.ConvertToDynamic(overwrite.Metadata);
+            overwriteObject.uid = overwrite.Uid;
+            var overwriteModel = new FileModel(owModel.FileAndType, overwriteObject, owModel.OriginalFileAndType);
+            var context = new ProcessContext(host, overwriteModel)
+            {
+                ContentAnchorParser = new ContentAnchorParser(overwrite.Conceptual)
+            };
+
+            var transformed = OverwriteProcessor.Process(overwriteObject, schema, context) as IDictionary<string, object>;
+            if (!context.ContentAnchorParser.ContainsAnchor)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs]: why this is a shared logic between overwrite markdown and markdown fragment?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs (hunk 1 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs]: `Reexport`

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs [hunk 4]
```diff
@@ -12,48 +12,71 @@ namespace Microsoft.DocAsCode.Build.SchemaDriven
     using Microsoft.DocAsCode.DataContracts.Common;
     using Microsoft.DocAsCode.Plugins;
 
-    public class ApplyOverwriteUtility
+    public class ApplyOverwriteHelper
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs]: naming: `Helper`, `Utility`, as convention, sounds like static classes?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteDocument.cs [hunk 5]
```diff
@@ -9,35 +9,20 @@ namespace Microsoft.DocAsCode.Build.SchemaDriven
...
-
-        private readonly Merger _merger = new Merger();
-
         public override string Name => nameof(ApplyOverwriteDocument);
 
         public override int BuildOrder => 0x10;
 
         public override void Postbuild(ImmutableList<FileModel> models, IHostService host)
         {
+            var ApplyOverwriteHelper = new ApplyOverwriteHelper(host, OverwriteModelType.OverwriteDocument);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteDocument.cs]: naming: lowercase

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            owModel.Properties.ExternalXRefSpecs = context.ExternalXRefSpecs;
+
+            foreach (var d in context.Dependency)
+            {
+                _host.ReportDependencyTo(owModel, d, DependencyTypeName.Include);
+            }
+            return transformed;
+        }
+
+        public static void MergeContentWithOverwrite(ref object source, object overwrite, string uid, string path, BaseSchema schema)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs]: some are public instance methods some are public static methods, what's the bar?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/OverwriteModelType.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.SchemaDriven
+{
+    public enum OverwriteModelType
+    {
+        OverwriteDocument,
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/Models/OverwriteModelType.cs]: naming: `Traditional`?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs (hunk 4 above)
Comment (928PJY) [reply src/Microsoft.DocAsCode.Build.SchemaDriven/ApplyOverwriteUtility.cs]: what about OverwriteApplier?

---
In reply to: [165292115](https://github.com/dotnet/docfx/pull/2398#discussion_r165292115) [](ancestors = 165292115)

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/OverwriteModelType.cs (hunk 7 above)
Comment (928PJY) [reply src/Microsoft.DocAsCode.Build.SchemaDriven/Models/OverwriteModelType.cs]: classic vs modern?

---
In reply to: [165293683](https://github.com/dotnet/docfx/pull/2398#discussion_r165293683) [](ancestors = 165293683)

Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Build.SchemaDriven/Models/OverwriteModelType.cs]: Classic?

**Pull request merged**